presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java

```diff
@@ -177,6 +178,10 @@ private boolean arePlansSame(FilterNode filter, TableScanNode tableScan, PlanNod
                             domainTranslator.toPredicate(newDomain.simplify().transform(assignments::get))));
             constraint = new Constraint(newDomain, evaluator::isCandidate);
         }
+        else if (newDomain.isAll() && !usesLegacyTableLayouts) {
```

Comment:
> By adding the check here, the flow is more confusing and harder to reason about.

True. I too independently added it here, because here we know whether we have evaluator....
but we can later check `contraint.predicate().isPresent()` too.

Reply:
Moved to `if (!metadata.usesLegacyTableLayouts(session, node.getTable())) {` branch